transportable-udfs-spark_2.11/src/main/scala/com/linkedin/transport/spark/StdUdfWrapper.scala

```diff
@@ -46,8 +49,7 @@ abstract class StdUdfWrapper(_expressions: Seq[Expression]) extends Expression
     _stdUdf.init(_stdFactory)
     getRequiredFiles()
     _requiredFilesProcessed = false
-    _outputDataType = sparkTypeInference.getOutputDataType
-    _outputDataType
+    sparkTypeInference.getOutputDataType
   }
```

Comment:
I think we should be able to refactor `initialize` so that most of it is called in `checkInputDataTypes` to initialize necessary members, and also still correctly implement `dataType`. Not sure if the `lazy` complexity is needed here.

Reply:
I think my current refactor already indicates this, the `checkInputDataTypes` will always evaluate `_outputDataType` which will then be lazily evaluated by `initialize`.

Reply:
Should not we be able to get rid of `lazy` and the explicit calls that force evaluating it?

Reply:
I think using lazy actually is cleaner, it saves us another state variable like `_requiredFilesProcessed` and some `if` condition branches.

Reply:
I think it saves `if (_outputDataType != null) _outputDataType`, which is good. `_requiredFilesProcessed` is still around though. Need to find out whether forcing it in `eval` is required. That would depend on whether `dataType` is being called on the driver or executor path. But even if it is called on the executor path, I think something like `makeCopy` should take care of sending driver state to the executor anyways.

Reply:
I'm not sure if the `makeCopy` really works, because the field is marked as `@transient` which means it can't be serialized. So I think the safest way is for the executor also to be able to construct such object from scratch and make sure the correct order. ie. `initialize` is called before `eval`